components/places/src/history_sync/payload_evolution_tests.rs

```diff
@@ -0,0 +1,336 @@
+            "visits": [
+                {
+                    "date": timestamp(1),
+                    "type": 1,
+                },
+                {
+                    "date": timestamp(2),
+                    "type": 1,
+                },
+            ]
+        })],
+        outgoing_unknown_fields: vec![OutgoingRecordUnknownFields {
+            guid: "a___________".into(),
+            fields: unknown_fields!({
+                "theme": "nord",
+            }),
+            visits_fields: vec![
+                (timestamp(1), unknown_fields!({})),
+                (timestamp(2), unknown_fields!({})),
+                (timestamp(10), unknown_fields!({})),
+            ],
+        }],
+    }
+    .run()
+}
+
+#[test]
+fn test_history_visits_roundtrip_fields() {
+    // Test that we roundtrip unknown fields from remote visits after making changes to them
+    RoundtripTest {
+        initial_remote_records: vec![json!({
+            "id": "a___________",
+            "title": "dogs",
+            "histUri": "https://dogs.com/",
+            "visits": [],
+            "visits": [
+                {
+                    "date": timestamp(1),
+                    "type": 1,
+                    // Some more fake new features for the tests:
+                    //   - What emotions did the page evoke?
+                    //   - How much fake news was listed?
+                    "emotion": "joy",
+                    "fake-news-amount": "some",
+                },
+                {
+                    "date": timestamp(2),
+                    "type": 1,
+                    "emotion": "anger",
+                },
+            ]
+        })],
+        local_visits: vec![visit("https://dogs.com/", timestamp(10))],
+        incoming_remote_records: vec![],
+        outgoing_unknown_fields: vec![OutgoingRecordUnknownFields {
+            guid: "a___________".into(),
+            fields: unknown_fields!({}),
+            visits_fields: vec![
+                (
+                    timestamp(1),
+                    unknown_fields!({
+                        "emotion": "joy",
+                        "fake-news-amount": "some",
+                    }),
+                ),
+                (
+                    timestamp(2),
+                    unknown_fields!({
+                        "emotion": "anger",
+                    }),
+                ),
+                (timestamp(10), unknown_fields!({})),
+            ],
+        }],
+    }
+    .run()
+}
+
+// The last case to consider is incoming visits with differerent unknown fields.  Maybe we should
+// handle that, but it would mean changing a lot of legacy code that assumes that visits are
+// immutable and can be compared based on the timestamp + transition fields (or the other code that
+// compares only the timestamps).  If we want to address this case, let's do it as part of a
+// general rework of the places sync merge code.
```

Comment:
What do others think about this one?  The current code treats visits are currently considered immutable, which means if we add unknown fields, they can't change either.

Should we try to rework the code so it doesn't make that assumption?  I kind of want to rework the merging code anyways and this could be a good excuse.

Another question is if we want to allow new fields to be part of the "key".  Visits don't have a GUID, we [treat two visits the same if they have the same timestamp + transition type](https://github.com/mozilla/application-services/blob/main/components/places/src/history_sync/plan.rs#L106-L122) (except [sometimes it's just timestamp](https://github.com/mozilla/application-services/blob/main/components/places/src/storage/history.rs#L921-L928) -- one more reason that we may want to rework this code).  Would we ever want to allow new fields to be added to this key?  Right now I don't think it's worth the complexity, but I wonder what others think.

Reply:
One thing I'm trying to keep in mind of roundtripping this data, is really to prevent data loss until the clients that can "catch up" and actually understand the data in those fields. 

Which means -- `unknown_fields` _should_ eventually disappear as a client becomes "recent", having unknown_fields stick around even if every client can understand the schema -- seems bad (IMO). I do like the idea of reworking so that we avoid this pitfall but depends on our timelines.

To your second question, I agree given the added complexity and we don't have a spec we'd be trying to design the would-be new "keys" around -- might be better to de-prioritize that.

Reply:
> Which means -- unknown_fields should eventually disappear as a client becomes "recent", having unknown_fields stick around even if every client can understand the schema -- seems bad (IMO). I do like the idea of reworking so that we avoid this pitfall but depends on our timelines.

I'm not understanding this in relation to this question.  I think we will eventually null out `unknown_fields` once we upgrade no matter what.  My question is if we should support changes to `unknown_fields` before that happens.

However, it's a really good point in general and it's making me rethink the pruning idea.

Reply:
I think a visit is considered immutable because it's recording a historical event - so changing it in the future really doesn't make sense. I can't see how a different device would somehow say "this old visit I previously recorded now should have different attributes".

Similarly, I think the key question is roughly the same - it's a record of something happening at an instant - I don't see how other fields being part of the key makes sense - you'd never want to record 2 different visits of the same type at the same timestamp with different attributes - there can only really have been one visit at that instant.

If that all makes sense, maybe just update the comment to reflect that?